test/pointcloud_voxelization_test.cpp

```diff
@@ -0,0 +1,257 @@
+  const double z_size = 2.0;
+  // 1/4 meter resolution, so 8 cells/axis
+  const double grid_resolution = 0.25;
+  const double step_size_multiplier = 0.5;
+  const CollisionCell empty_cell(0.0f);
+  const CollisionCell filled_cell(1.0f);
+
+  const common_robotics_utilities::voxel_grid::GridSizes grid_size(
+      grid_resolution, x_size, y_size, z_size);
+  CollisionMap static_environment(X_WG, "world", grid_size, empty_cell);
+
+  // Set the bottom cells filled
+  for (int64_t xidx = 0; xidx < static_environment.GetNumXCells(); xidx++)
+  {
+    for (int64_t yidx = 0; yidx < static_environment.GetNumYCells(); yidx++)
+    {
+      static_environment.SetValue(xidx, yidx, 0, filled_cell);
+    }
+  }
+
+  // Make some test pointclouds
+  // Make the physical->optical frame transform
+  const Eigen::Isometry3d X_CO = Eigen::Translation3d(0.0, 0.0, 0.0) *
+      Eigen::Quaterniond(Eigen::AngleAxisd(-M_PI_2, Eigen::Vector3d::UnitZ()) *
+                         Eigen::AngleAxisd(-M_PI_2, Eigen::Vector3d::UnitX()));
+
+  // Camera 1 pose
+  const Eigen::Isometry3d X_WC1(Eigen::Translation3d(-2.0, 0.0, 0.0));
+  const Eigen::Isometry3d X_WC1O = X_WC1 * X_CO;
+  PointCloudWrapperPtr cam1_cloud(new VectorVector3dPointCloudWrapper());
+  static_cast<VectorVector3dPointCloudWrapper*>(cam1_cloud.get())
+      ->SetPointCloudOriginTransform(X_WC1O);
+  for (double x = -2.0; x <= 2.0; x += 0.03125)
+  {
+    for (double y = -2.0; y <= 2.0; y += 0.03125)
+    {
+      const double z = (x <= 0.0) ? 2.125 : 4.0;
+      static_cast<VectorVector3dPointCloudWrapper*>(cam1_cloud.get())->PushBack(
+          Eigen::Vector3d(x, y, z));
+    }
+  }
+
+  // Camera 2 pose
+  const Eigen::Isometry3d X_WC2 = Eigen::Translation3d(0.0, -2.0, 0.0) *
+      Eigen::Quaterniond(Eigen::AngleAxisd(M_PI_2, Eigen::Vector3d::UnitZ()));
+  const Eigen::Isometry3d X_WC2O = X_WC2 * X_CO;
+  PointCloudWrapperPtr cam2_cloud(new VectorVector3dPointCloudWrapper());
+  static_cast<VectorVector3dPointCloudWrapper*>(cam2_cloud.get())
+      ->SetPointCloudOriginTransform(X_WC2O);
+  for (double x = -2.0; x <= 2.0; x += 0.03125)
+  {
+    for (double y = -2.0; y <= 2.0; y += 0.03125)
+    {
+      const double z = (x >= 0.0) ? 2.125 : 4.0;
+      static_cast<VectorVector3dPointCloudWrapper*>(cam2_cloud.get())->PushBack(
+          Eigen::Vector3d(x, y, z));
+    }
+  }
+
+  // Make control parameters
+  // We require that 100% of points from the camera see through to see a voxel
+  // as free.
+  const double percent_seen_free = 1.0;
+  // We don't worry about outliers.
+  const int32_t outlier_points_threshold = 1;
+  // We only need one camera to see a voxel as free.
+  const int32_t num_cameras_seen_free = 1;
+  const PointCloudVoxelizationFilterOptions filter_options(
+      percent_seen_free, outlier_points_threshold, num_cameras_seen_free);
+  // Voxelizer options (leave as default).
+  const std::map<std::string, int32_t> options;
+
+  // CUDA voxelizer
+  try
```

Comment:
@calderpg nit: putting CUDA, OpenCL, and CPU voxelization in separate tests makes it easier to pinpoint issues. That means all setup code immediately above has to be relocated to a fixture class.

Reply:
So far they've lived in a single test mostly because you don't know which of the three variants will actually run on a given platform, and failure of the Cuda and OpenCL variants to run doesn't mean the test should fail.